### DISCUSSION TITLE
Integrate with custom GPT for "asking the docs"

### DISCUSSION THREAD

**Discussion by otisg** [mkdocs/mkdocs#3582] [Ideas]
I think this would be a killer feature.
Allow one to integrate their mkdocs with e.g. OpenAI where one can build a custom GPT (e.g. from docs), so that in addition to the regular mkdocs search -- or maybe in place of is -- one can prompt/query the docs using the custom GPT behind the scenes.

**Comment by otisg** [comment]
@oprypin this already exists?

**Comment by pawamoy** [comment]
I've seen some examples, yes: https://pdm-project.org/latest/ (enable all Javascript and click on the robot head in the bottom right corner). The sources for the chat window seem to be here: https://github.com/pdm-project/pdm/blob/main/chatbot/app.py. 

There's also a chat bot in FastAPI's docs: https://fastapi.tiangolo.com/.

Let me move this to a discussion.

**Reply by pawamoy** [reply]
Actually the PDM chat bot does not seem to work on my end. There's no input and the fullscreen app just shows "Running...".

**Reply by pawamoy** [reply]
Also, code that integrates the streamlit-based chatbot in the docs: https://github.com/pdm-project/pdm/blob/main/docs/overrides/main.html.

**Comment by otisg** [comment]
Thanks for these links. OK, so *somebody* has created a chatbot and put it on the mkdocs-generated website... but that's not the same as having this functionality as part of mkdocs. What am I missing?

**Reply by pawamoy** [reply]
I don't think such functionality will ever be integrated into MkDocs itself. 

- The original suggestion is too vague
- Every site would probably have specific needs that a single solution will not be able to cater for
- I'd say ChatBots are out-of-scope for MkDocs itself (do one thing and do it well)
- The ChatBot has to run somewhere, and MkDocs generates static sites
- It would be better written as a MkDocs plugin + other tools to help with setup/configuration/deployment

Now that is just my own opinion and maybe other maintainers will have different ones :slightly_smiling_face:

**Reply by gbathree** [reply]
hmmmm... I'm not sure if that answer will hold up well in a year or two.

**Reply by pawamoy** [reply]
I think it will, unless MkDocs starts being a HTTP server or something :)

**Reply by pawamoy** [reply]
Do note that I'm not arguing against the feature itself. I'm arguing against implementing it in MkDocs itself. This is too complex a subject for a static site generator to provide a single solution for it. Themes, frameworks, plugins are welcome to try and provide a solution :)

**Reply by gbathree** [reply]
I get it, that's fair.  But from a user's point of view, mkdocs today has search which is expected and is a core feature.  Users similarly will expect a better search (which will be in some way llm supported) in the future.

So while from a technical perspective this may seem separate, from an actual users perspective, they will absolutely expect llm based search interactions in the future as core.

**Reply by pawamoy** [reply]
Yeah that's a good point :)

**Reply by scottdunnflux** [reply]
> I get it, that's fair. But from a user's point of view, mkdocs today has search which is expected and is a core feature. Users similarly will expect a better search (which will be in some way llm supported) in the future.
> 
> So while from a technical perspective this may seem separate, from an actual users perspective, they will absolutely expect llm based search interactions in the future as core.

Two things
1. my markdown.yml file shows 
    ```yaml
    plugins:
      - search
    ```
    So, it feels like a plugin ;-) —one that could be disabled if you don't need it, want to take a different approach, or want a lighter site build.
2. The mkdocs search plugin is great because the final site will work without a server or an internet connection. Perhaps, someday soon, there will be a way to put a good lightweight LLM inside a browser-run JavaScript that might be tuned or include a vector database generated during the `mkdocs build` process.

Until then I wouldn't mind a plug-in that would ask for API keys to hosted vector databases and LLMs and automate the process of creating the databases and adding the javascript for the chat interface to some or all pages in a site.

**Comment by Andre601** [comment]
This feels like something not suited for MkDocs.

MkDocs is for building static sites, not integrating some custom Chat bot/AI system into itself. That said, the plugin API exists and one could just make a plugin for this feature.

Also, you can (most likely) customize a theme to integrate this functionality into the final site...
